sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+            state_idx += layer.num_states(step)
+        else:
+            layer_states = []
+
+        yield layer, layer_states
+
+
+class NestedDecoderLayer(DecoderLayer):
+    """
+    A decoder layer which combines several other decoder sub-layers.
+    """
+
+    @property
+    @abstractmethod
+    def layers(self) -> List[DecoderLayer]:
+        pass
+
+    def layers_with_states_iter(self,
+                                step: int,
+                                layer_states_flat: List[mx.sym.Symbol]) -> Iterator[Tuple[DecoderLayer,
+                                                                                          List[mx.sym.Symbol]]]:
+        return layers_with_states_iter(self.layers, step, layer_states_flat)
+
+    def att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.att_names())
+        return att_names
+
+    def self_att_names(self):
+        att_names = []
+        for layer in self.layers:
+            att_names.extend(layer.self_att_names())
+        return att_names
+
+    def reset(self):
+        for layer in self.layers:
+            layer.reset()
+
+    def num_states(self, step):
+        return sum(layer.num_states(step) for layer in self.layers)
+
+    def state_variables(self, step: int):
+        return [var for layer in self.layers for var in layer.state_variables(step)]
+
+    def state_shapes(self,
+                     batch_size: int,
+                     step: int,
+                     source_encoded_max_length: int,
+                     source_encoded_num_hidden: int):
+        return [state_shape for layer in self.layers for state_shape in layer.state_shapes(batch_size,
+                                                                                           step,
+                                                                                           source_encoded_max_length,
+                                                                                           source_encoded_num_hidden)]
+
+    def init_states(self,
+                    batch_size: int,
+                    source_encoded: List[mx.sym.Symbol],
+                    source_encoded_lengths: mx.sym.Symbol,
+                    source_encoded_max_length: int):
+        return [init_state for layer in self.layers for init_state in layer.init_states(batch_size,
+                                                                                        source_encoded,
+                                                                                        source_encoded_lengths,
+                                                                                        source_encoded_max_length)]
+
+    def get_max_seq_len(self):
+        return min((layer.get_max_seq_len() for layer in self.layers if layer.get_max_seq_len() is not None),
+                   default=None)
+
+
+class EncoderLayerChain(EncoderLayer):
```

Comment:
docstring.
Also, why are DecoderLayers 'nested' and encoder layers 'chained'?

Reply:
While `NestedDecoderLayer` is meant to have all the state handling logic when you have multiple sub-layers (where it is undefined in which way the sub-layers are used), the layer chain will pass a hidden state through a chain of layers. The DecoderLayerChain then subclasses the NestedDecoderLayer. On the encoder side there is no state handling so that there is no corresponding NestedDecoderLayer.

Reply:
That said *LayerChain classes might be an unused artifact that we can potentially remove.